x/exchange/keeper/grpc_query.go

```diff
@@ -545,3 +545,172 @@
+		source, externalID, pErr := ParseIndexKeySuffixTargetToPayment(keySuffix)
+		if pErr != nil {
+			k.logEndpointError(ctx, "GetPaymentsWithTarget", "Error reading target to payment index entry.",
+				"error", pErr, "target", target.String(),
+				"keyPrefix", fmt.Sprintf("%v", keyPrefix), "keySuffix", fmt.Sprintf("%v", keySuffix))
+			return nil
+		}
+
+		payment, pErr := k.getPaymentFromStore(store, source, externalID)
+		if pErr != nil {
+			k.logEndpointError(ctx, "GetPaymentsWithTarget", "Error reading payment from store.", "error", pErr,
+				"target", target.String(), "source", source.String(), "externalID", externalID,
+			)
+			return nil
+		}
+		if payment == nil {
+			k.logEndpointError(ctx, "GetPaymentsWithTarget", "No payment found from target to payment index entry.",
+				"target", target.String(), "source", source.String(), "externalID", externalID)
+			return nil
+		}
+
+		resp.Payments = append(resp.Payments, payment)
+		return nil
+	})
+
+	if pageErr != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "error iterating payments with target %s: %v", req.Target, pageErr)
+	}
+
+	return resp, nil
+}
+
+// GetAllPayments gets all payments.
+func (k QueryServer) GetAllPayments(goCtx context.Context, req *exchange.QueryGetAllPaymentsRequest) (*exchange.QueryGetAllPaymentsResponse, error) {
+	var pagination *query.PageRequest
+	if req != nil {
+		pagination = req.Pagination
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	keyPrefix := GetKeyPrefixAllPayments()
+	preStore := prefix.NewStore(k.getStore(ctx), keyPrefix)
+
+	resp := &exchange.QueryGetAllPaymentsResponse{}
+	var pageErr error
+	resp.Pagination, pageErr = query.Paginate(preStore, pagination, func(keySuffix, value []byte) error {
+		payment, pErr := k.parsePaymentStoreValue(value)
+		if pErr != nil {
+			k.logEndpointError(ctx, "GetAllPayments", "Error reading payment from store.",
+				"error", pErr, "value", fmt.Sprintf("%v", value),
+				"keyPrefix", fmt.Sprintf("%v", keyPrefix), "keySuffix", fmt.Sprintf("%v", keySuffix))
+			return nil
+		}
+		if payment == nil {
+			k.logEndpointError(ctx, "GetAllPayments", "Empty payment entry.",
+				"value", fmt.Sprintf("%v", value),
+				"keyPrefix", fmt.Sprintf("%v", keyPrefix), "keySuffix", fmt.Sprintf("%v", keySuffix))
+			return nil
+		}
+		resp.Payments = append(resp.Payments, payment)
+		return nil
+	})
+
+	if pageErr != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "error iterating all payments: %v", pageErr)
+	}
+
+	return resp, nil
+}
+
+// PaymentFeeCalc calculates the fees that must be paid for creating or accepting a specific payment.
+func (k QueryServer) PaymentFeeCalc(goCtx context.Context, req *exchange.QueryPaymentFeeCalcRequest) (*exchange.QueryPaymentFeeCalcResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	resp := k.CalculatePaymentFees(ctx, &req.Payment)
+	return resp, nil
+}
```

Comment:
Really, for this calculation, all that's needed is the `source_amount` and/or `target_amount`. The reason `QueryPaymentFeeCalcRequest` takes in a `Payment` is so that users can provide the same thing here that they'll provide to either the `CreatePayment` or `AcceptPayment` endpoints. Those endpoints are where the meaningful validation is done.

Reply:
The `PaymentFeeCalc` endpoint already does all the validation that it needs to do. It's valid for either `source_amount` or `target_amount` to be nil or empty, and the actual entries in those fields don't matter for the calculation.